src/timely-util/src/progress.rs

```diff
@@ -27,3 +30,392 @@ where
+/// ## The problem of partitioned timestamps
+///
+/// To understand this timestamp it helps to first think about how one would represent a
+/// partitioned, partially ordered timestamp type where each partition follows its own independent
+/// timeline in a natural way. Timely requires that all timestamps have a minimum element that is
+/// less than all other elements so you could express a partitioned timestamp like so:
+///
+/// ```rust
+/// enum Partitioned<P, T> {
+///     Bottom,
+///     Partition(P, T),
+/// }
+/// ```
+/// Laying out the order in a graph we'd see something like this:
+///
+/// ```text
+///           -----(P1, T0)---(P1, T1)--- ... ---(P1, Tn)
+///          / ----(P2, T0)---(P2, T1)--- ... ---(P2, Tn)
+///         / / ,--(P3, T0)---(P3, T1)--- ... ---(P3, Tn)
+/// Bottom-+-+-+     .
+///         \        .
+///          \       .
+///           '----(Pn, T0)---(Pn, T1)--- ... ---(Pn, Tn)
+/// ```
+///
+/// This timestamp has the problem that if you want to downgrade your operator's capability to
+/// indicate progress in one of the partitions, which implies dropping your `Bottom` capability,
+/// you are forced to also mint one capability per partition that you intend to produce data for in
+/// the future. If those partitions are unknown or, even worse, infinite, this type brings you to a
+/// dead end.
+///
+/// ## How this type works
+///
+/// The idea of this `Parititoned` timestamp is to extend the idea of the `Bottom` element above
+/// into a `Range` element that is parameterized by a lower and upper bound and represents a
+/// *range* of partitions at some timestamp T. The represented range has exclusive bounds.
+///
+/// The minimum timestamp of this type is `Range(Bottom, Bottom, T::minimum(0)` which is less than
+/// any other `Range` element and all `Partition` elements. Now, suppose an operator needs to start
+/// working on some partition `P1` and present progress. All it has to do is downgrade its
+/// `Antichain { Range(Bottom, Top, 0) }` frontier in this frontier:
+/// `Antichain { Range(Bottom, Elem(P1), 0), Partition(P1, T::minimum()), Range(Elem(P1), Top, 0) }`.
+///
+/// Essentially a `Range` element can be split at some partition P iff that partition is within
+/// its range and produce two more `Range` elements representing the range to the left and to the
+/// right of the partition respectively, plus a timestamp for the desired partition that can now be
+/// downgraded individually to present progress.
+///
+/// Antichains of this type are efficient in storage. In the worst case, where all chosen
+/// partitions have gaps between them, the produced antichain has twice as many elements as
+/// partitions.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+pub enum Partitioned<P, T> {
+    /// A Range element representing an exclusive range of partitions.
+    Range(RangeBound<P>, RangeBound<P>, T),
+    /// Timestamp T for partition P
+    Partition(P, T),
+}
```

Comment:
It seems like a lot of this could be simplified if you had an 
```rust
enum Interval<P> { 
    Range(RangeBound<P>, RangeBound<P>),
    Point(P),
}
```
as most of the rest of the logic doesn't seem to take specific `T`-action based on the type of the interval.

Said differently, I think a lot of this is a compressed representation of `Map<P, T>`, atomized into individual parts. The interesting logic is how the atoms implement `Map` rather than their relationship to `T`. You could almost even implement this with `Product<Interval<P>, T>`, except for what seem like some quirks about non-intersection (perhaps you could if the summary type was `Map<P, T::Summary>` rather than the atoms).

Reply:
I just had time to look into this again. This was a great idea! I changed the PR to just wrap `Product` and all tests passed without modification :)